### ISSUE TITLE
Broadcasting Error in production

### ISSUE THREAD

**Issue by alliedarmour** [joshleblanc/view_component_reflex#9]
In development the gem is working great, the other Stimulus Reflex stuff is working aswell in production - but If I try to use View Component Reflex in production I get the following error:

`StimulusReflex::Channel#receive({"target"=>"PerformedOrdersComponent#get_orders", "args"=>[], "url"=>"https://webmenue.nkbrf.de/users/9/show_orders", "attrs"=>{"id"=>"date_month", "name"=>"date[month]", "class"=>"form-control shadow font-weight-bold", "data-reflex"=>"change->PerformedOrdersComponent#get_orders", "data-user-id"=>"9", "data-key"=>"4002470145471996799", "data-controller"=>"stimulus-reflex", "data-action"=>"change->stimulus-reflex#__perform", "checked"=>false, "selected"=>false, "tag_name"=>"SELECT", "values"=>["6"], "value"=>"6"}, "dataset"=>{"data-reflex"=>"change->PerformedOrdersComponent#get_orders", "data-user-id"=>"9", "data-key"=>"4002470145471996799", "data-controller"=>"stimulus-reflex", "data-action"=>"change->stimulus-reflex#__perform"}, "selectors"=>[], "reflexId"=>"9474f0a5-f262-499a-bbd7-88defe71f477", "permanent_attribute_name"=>"data-reflex-permanent", "params"=>{}})`

`Could not execute command from ({"command"=>"message", "identifier"=>"{\"channel\":\"StimulusReflex::Channel\"}", "data"=>"{\"target\":\"PerformedOrdersComponent#get_orders\",\"args\":[],\"url\":\"https://webmenue.nkbrf.de/users/9/show_orders\",\"attrs\":{\"id\":\"date_month\",\"name\":\"date[month]\",\"class\":\"form-control shadow font-weight-bold\",\"data-reflex\":\"change->PerformedOrdersComponent#get_orders\",\"data-user-id\":\"9\",\"data-key\":\"4002470145471996799\",\"data-controller\":\"stimulus-reflex\",\"data-action\":\"change->stimulus-reflex#__perform\",\"checked\":false,\"selected\":false,\"tag_name\":\"SELECT\",\"values\":[\"6\"],\"value\":\"6\"},\"dataset\":{\"data-reflex\":\"change->PerformedOrdersComponent#get_orders\",\"data-user-id\":\"9\",\"data-key\":\"4002470145471996799\",\"data-controller\":\"stimulus-reflex\",\"data-action\":\"change->stimulus-reflex#__perform\"},\"selectors\":[],\"reflexId\":\"9474f0a5-f262-499a-bbd7-88defe71f477\",\"permanent_attribute_name\":\"data-reflex-permanent\",\"params\":{}}"}) [NoMethodError - undefined method `broadcast_message' for nil:NilClass]: /home/deploy/webmenue/shared/bundle/ruby/2.6.0/gems/stimulus_reflex-3.3.0.pre2/lib/stimulus_reflex/channel.rb:37:in `rescue in receive' | /home/deploy/webmenue/shared/bundle/ruby/2.6.0/gems/stimulus_reflex-3.3.0.pre2/lib/stimulus_reflex/channel.rb:30:in `receive' | /home/deploy/webmenue/shared/bundle/ruby/2.6.0/gems/actioncable-6.0.3.2/lib/action_cable/channel/base.rb:268:in `public_send' | /home/deploy/webmenue/shared/bundle/ruby/2.6.0/gems/actioncable-6.0.3.2/lib/action_cable/channel/base.rb:268:in `dispatch_action' | /home/deploy/webmenue/shared/bundle/ruby/2.6.0/gems/actioncable-6.0.3.2/lib/action_cable/channel/base.rb:170:in `block in perform_action'`

The channel is working normally:

`StimulusReflex::Channel is streaming from StimulusReflex::Channel:9
`
It's just happening in production, any idea?

**Comment by joshleblanc** [comment]
That error implies that the reflex class wasn't found. Someone else ran into this in the discord but I couldn't reproduce it.

Can you create an MVCE so I can look at this on my end?

**Comment by alliedarmour** [comment]
> That error implies that the reflex class wasn't found. Someone else ran into this in the discord but I couldn't reproduce it.
> 
> Can you create an MVCE so I can look at this on my end?

Mh, as it's just not working in production that could be a bit hard. 
I can provide some snippets of my code, maybe this helps to figure out the problem

Basically I have a normal Rails view which is rendering the component:

```
<div class="container-fluid">
    <h3>Ihre Bestellungen für den Monat ...</h3>
    <%= render(PerformedOrdersComponent.new(orders: @orders, user: @user)) %>
</div>
```
Next up, inside the component I'm initializing the data and defining my reflex:

```
class PerformedOrdersComponent < ViewComponentReflex::Component
    extend Forwardable

    attr_reader :orders, :user

    def_delegator :@user, :id, :user_id
    
    def initialize(orders:, user:)
        @orders ||= orders
        @user = user
    end

    def get_orders
        month = element.value.to_i
        user = User.find(element.dataset.user_id.to_i)
        @orders = user.order_cards.where(day: Date.new(Date.today.year, month).all_month).order(day: :asc)
    end
end
```

And this is my view template:

```
<%= component_controller do %>
    <%= select_month(Date.today, { use_month_names: %w(Januar Februar März April Mai Juni Juli August September Oktober November Dezember) }, 
    class: "form-control shadow font-weight-bold", data: { reflex: "change->PerformedOrdersComponent#get_orders", user_id: user_id, key: key } ) %>

    <hr style="width: 100%; color: black; height: 1px; background-color:black;"/>
        
    <div class="table-responsive">
        <table class="table table-borderless table-striped table-condensed shadow">
            <thead class="bg-bdh text-white text-center">
                <tr>
                    <th>Datum</th>
                    <th>Mahlzeit</th>
                    <th>Preis</th>
                </tr>
            </thead>
            <tbody>
                <%= render(OrderComponent.with_collection(orders, data: { key: key })) %>                      
            </tbody>
            <tfooter>
                <% if orders.present? %>
                    <td colspan="2" class="bg-white"></td>
                    <td class="bg-bdh text-white">
                        <div class="h5 font-weight-bold">
                            <%= number_to_currency(orders.sum(:price), unit: "€", separator: ",", format: "%n%u") %>
                        </div>
                    </td>
                <% else %>
                    <td colspan="3" class="bg-white">
                        <div class="alert alert-danger">Für den ausgewählten Monat sind keine Bestellungen vorhanden.</div>
                    </td>
                <% end %>
                </tr>         
            </tfooter>
        </table>
    </div>
<% end %>
```

And inside I'm rendering a component for each order, I tried it with normal View component aswell as with the reflex view component by passing the key, but no luck :/:

```
class OrderComponent < ViewComponentReflex::Component
  extend Forwardable

  def_delegators :@order, :day, :dish_type, :price

  def initialize(order:, data: {})
    @order = order
  end

  def collection_key
    @order.id
  end
end
```

As I said they both work in development, but not on production server. Did I miss a component_controller or something?

**Comment by joshleblanc** [comment]
No, if it works in development, it should work in production. 

I wonder if there's some autoloading bug going on. The reflexes are created at runtime (the first time component_controller is called actually), so there's no reason it that error *could* happen, especially since the expo site doesn't have this problem.


I'll try using the code samples you provided to re-create it after work

**Comment by joshleblanc** [comment]
@alliedarmour What does order_component.html.erb look like?

**Comment by alliedarmour** [comment]
> @alliedarmour What does order_component.html.erb look like?

The order_component.html.erb just represents the rows of the table:

```
<tr>
    <td><%= I18n.l(day, format: :long) %></td>
    <td><%= dish_type %></td>
    <td><%= number_to_currency(price, unit: "€", separator: ",", format: "%n%u") %></td>
</tr>
```

I don't really think it's connected to my components, as I read in the discord that for the other guy it didn't work with the barebones example of incrementing a number.
I also checked redis, my cable config etc. but everything else is working fine, including regular StimulusReflex. 

Maybe problems with passenger or something? What's your expo running on?

**Comment by joshleblanc** [comment]
I have 2 sites running this on puma. I'll try launching it with passenger to see if that's the problem.

I've already ruled out redis as the problem

**Comment by joshleblanc** [comment]
I just straight cloned the expo and pushed it to heroku and it worked without any issues. 

This is very confusing.

**Comment by alliedarmour** [comment]
> I just straight cloned the expo and pushed it to heroku and it worked without any issues.
> 
> This is very confusing.

Mh, I have two applications running on Nginx + Passenger and SR running without any issues. Your gem is awesome, that's why I kinda don't like that I had to switch back to "regular" Reflexes, I'd rather had the logic in my component.

Maybe really some [autoloading](https://dev.to/shushugah/rails-auto-loading-magic-gone-wrong-19kp) issue?

**Comment by joshleblanc** [comment]
I'll try changing everything to thier full module paths after work

**Comment by joshleblanc** [comment]
I went ahead and pushed that change

**Comment by alliedarmour** [comment]
Hey, first thanks for your effort! Now at least the transmission seems to be working, but this time I get an javascript error in console:

![Bildschirmfoto von 2020-08-27 07-23-35](https://user-images.githubusercontent.com/20596363/91388015-cdfcdb80-e836-11ea-9754-80df7a46e08b.png)

The production log file says the following:

```
I, [2020-08-27T07:19:08.606411 #18771]  INFO -- : StimulusReflex::Channel#receive({"target"=>"PerformedOrdersComponent#increment", "args"=>[], "url"=>"https://webmenue.nkbrf.de/users/9/show_orders", "attrs"=>{"data-reflex"=>"click->PerformedOrdersComponent#increment", "data-key"=>"1297763082027554745", "data-controller"=>"stimulus-reflex", "data-action"=>"click->stimulus-reflex#__perform", "checked"=>false, "selected"=>false, "tag_name"=>"BUTTON", "value"=>""}, "dataset"=>{"data-reflex"=>"click->PerformedOrdersComponent#increment", "data-key"=>"1297763082027554745", "data-controller"=>"stimulus-reflex", "data-action"=>"click->stimulus-reflex#__perform"}, "selectors"=>[], "reflexId"=>"3eb0defe-e93b-41d8-a183-fc4128f38136", "permanent_attribute_name"=>"data-reflex-permanent", "params"=>{}})
E, [2020-08-27T07:19:08.607741 #18771] ERROR -- : [31m**StimulusReflex::Channel Failed to invoke PerformedOrdersComponent#increment! https://webmenue.nkbrf.de/users/9/show_orders undefined method `+' for nil:NilClass /home/deploy/webmenue/releases/20200827051441/app/components/performed_orders_component.rb:15:in `increment'**[0m
D, [2020-08-27T07:19:08.608129 #18771] DEBUG -- : [ActionCable] Broadcasting to StimulusReflex::Channel:9: {"cableReady"=>true, "operations"=>{"dispatchEvent"=>[{"name"=>"stimulus-reflex:server-message", "detail"=>{"stimulusReflex"=>{"target"=>"PerformedOrdersComponent#increment", "args"=>[], "url"=>"https://webmenue.nkbrf.de/users/9/show_orders", "attrs"=>{"data-reflex"=>"click->PerformedOrdersComponent#increment", "data-key"=>"1297763082027554745", "data-controller"=>"stimulus-reflex", "data-action"=>"click->stimulus-reflex#__perform", "checked"=>false, "selected"=>false, "tagName"=>"BUTTON", "value"=>""}, "dataset"=>{"data-reflex"=>"click->PerformedOrdersComponent#increment", "data-key"=>"1297763082027554745", "data-controller"=>"stimulus-reflex", "data-action"=>"click->stimulus-reflex#__perform"}, "selectors"=>["body"], "reflexId"=>"3eb0defe-e93b-41d8-a183-fc4128f38136", "permanentAttributeName"=>"data-reflex-permanent", "params"=>{}, "broadcaster"=>:nothing, "serverMessage"=>{"subject"=>"error", "body"=>"StimulusReflex::Channel Failed to invoke PerformedOrdersComponent#increment! https://webmenue.nkbrf.de/users/9/show_orders undefined method `+' for nil:NilClass /home/deploy/webmenue/releases/20200827051441/app/components/performed_orders_component.rb:15:in `increment'"}}}}]}}
D, [2020-08-27T07:19:08.609510 #18771] DEBUG -- : StimulusReflex::Channel transmitting {"cableReady"=>true, "operations"=>{"dispatchEvent"=>[{"name"=>"stimulus-reflex:server-message", "detail"=>{"stimulusReflex"=>{"target"=>"PerformedOrdersComponent#increment", "args"=>[], "url"=>"https://webmenue.nkbrf.de/users/9/show_orders", "attrs"=>{"data-reflex"=>"click->PerformedOrdersCompon... (via streamed from StimulusReflex::Channel:9)

```

**Comment by joshleblanc** [comment]
This is a different error, I'd need to see the component code and markup for this particular case.

**Comment by alliedarmour** [comment]
> This is a different error, I'd need to see the component code and markup for this particular case.
> 
> Are you running multiple production servers behind a load balancer by any chance?

It's just the standard increment code, just with another component name.

```
# performed_orders_component.rb
class PerformedOrdersComponent< ViewComponentReflex::Component
  def initialize
    @count = 0
  end

  def increment
    @count += 1
  end
end
```

```
# performed_orders_component.html.rb
<%= component_controller do %>
    <p><%= @count %></p>
    <%= reflex_tag :increment, :button, "Click" %>
<% end %>
```
Nope, it's just the standard capistrano / passenger / nginx deployment involved. I don't know may be something is wrong on my side and the error is not on your side, just wondering that SR itself works without problems. Just wondering that I'm not the only one with this problem.

**Comment by joshleblanc** [comment]
Very strange. The error you're getting is because `@count` is undefined. 

The way view_component_reflex works is when you render the page, it gathers all of the instance variables and tucks them away. When you perform a reflex, view_component_reflex re-creates the component, and sets all of the instance variables based on what it tucked away. Then it performs the reflex, tucks away the new variables, and renders the component.

So what's happening here is it's failing to re-initialize the component properly for some reason.

What happens if you set 

```ruby
  config.session_store :cache_store
```

In `production.rb`?

**Comment by alliedarmour** [comment]
> What happens if you set
> 
> ```ruby
>   config.session_store :cache_store
> ```
> 
> In `production.rb`?

This gives me the old error again x). Maybe I'll try to deploy to Heroku to see if it runs there.

**Comment by joshleblanc** [comment]
I'm confounded. I wish I could reproduce this

**Comment by joshleblanc** [comment]
That has been resolved in 2.3.1

**Comment by alliedarmour** [comment]
> That has been resolved in 2.3.1

Thanks for the reply! I can confirm it's working for me now :).
